source/Cosmos.IL2CPU/IL/Conv_I2.cs

```diff
@@ -24,48 +21,45 @@ public override void Execute(_MethodInfo aMethod, ILOpCode aOpCode)
             var xSource = aOpCode.StackPopTypes[0];
             var xSourceSize = SizeOfType(xSource);
             var xSourceIsFloat = TypeIsFloat(xSource);
-            if (xSourceIsFloat)
+
+            if (xSourceSize > 8)
+            {
+                throw new NotImplementedException("Cosmos.IL2CPU.x86->IL->Conv_I2.cs->Error: StackSize > 8 not supported");
+            }
+
+            if (xSourceSize <= 4)
             {
-                if (xSourceSize == 4)
+                if (xSourceIsFloat)
                 {
                     XS.SSE.MoveSS(XMM0, ESP, sourceIsIndirect: true);
                     XS.SSE.ConvertSS2SIAndTruncate(EAX, XMM0);
+                    XS.MoveSignExtend(EAX, AX);
                     XS.Set(ESP, EAX, destinationIsIndirect: true);
                 }
-                else if (xSourceSize == 8)
+                else
+                {
+                    XS.Pop(EAX);
+                    XS.MoveSignExtend(EAX, AX);
+                    XS.Push(EAX);
+                }
+            }
+            else if (xSourceSize <= 8)
+            {
+                if (xSourceIsFloat)
                 {
                     XS.SSE2.MoveSD(XMM0, ESP, sourceIsIndirect: true);
+                    XS.Add(ESP, 4);
                     XS.SSE2.ConvertSD2SIAndTruncate(EAX, XMM0);
                     XS.Set(ESP, EAX, destinationIsIndirect: true);
                 }

```

Comment:
I preferred the switch here sincerely if the intent of 'xSourceSize <= 4' was that sizes 1 (byte) and 2 (short) should have been comprised I preferred this:

```
case 1:
case 2:
case 4:
   [...]
```
So you avoid to accept size 3 by accident...

 else if (xSourceSize <= 8) is yet less clear we want accept really size 5, 6 and 7? I think not...

IMHO the ideal as now C# 7 support will be to use type pattern matching in all these case so we avoid the annoying xSourceIsFloat tests too:

https://visualstudiomagazine.com/articles/2017/02/01/pattern-matching.aspx